Bottom: 1ebcc46b30053d0cb464e4981eaeb79423b9ab04
Top:    912e4fa17fce1dbafff5d68dc6b7f40f0d02ec42
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-12-04 19:24:34 +0000

Fixup undo in line clearance check code


---

diff --git a/src/action.c b/src/action.c
index e7a2ea4..4e8ceb7 100644
--- a/src/action.c
+++ b/src/action.c
@@ -900,6 +900,7 @@ NotifyLine (void)
                                    &ptr3);
               LookupConnection (Crosshair.X, Crosshair.Y, true, 1, CONNECTEDFLAG, false, true);
               LookupConnection (Crosshair.X, Crosshair.Y, true, 1, FOUNDFLAG, true, true);
+              /* XXX: What about undo serial number? */
             }
           /* XXX: NEED TO FIGURE OUT WHAT NET CLASS THIS IS, AND/OR STORE FOR USE WITH DRC */
 	  Crosshair.Netclass = get_netclass_at_xy (LAYER_ON_STACK(0), Crosshair.X, Crosshair.Y); /* XXX: Not sure about the layer! */
diff --git a/src/find.c b/src/find.c
index 655d227..01e815f 100644
--- a/src/find.c
+++ b/src/find.c
@@ -3073,7 +3073,7 @@ LookupConnection (Coord X, Coord Y, bool AndDraw, Coord Range, int flag,
         LOOKUP_MORE & ~(AndRats ? 0 : RATLINE_TYPE),
         &ptr1, &ptr2, &ptr3, X, Y, Range);
       if (type == NO_TYPE)
-        return;
+        goto out;
       if (type & SILK_TYPE)
         {
           int laynum = GetLayerNumber (PCB->Data,
@@ -3081,12 +3081,16 @@ LookupConnection (Coord X, Coord Y, bool AndDraw, Coord Range, int flag,
 
           /* don't mess with non-conducting objects! */
           if (laynum >= max_copper_layer || ((LayerType *)ptr1)->no_drc)
-            return;
+            goto out;
         }
     }
 
-  name = ConnectionName (type, ptr1, ptr2);
-  hid_actionl ("NetlistShow", name, NULL);
+
+  if (AndDraw) /* Skip highlighting the net if we're not drawing */
+    {
+      name = ConnectionName (type, ptr1, ptr2);
+      hid_actionl ("NetlistShow", name, NULL);
+    }
 
   User = store_undo;
   InitConnectionLookup ();
@@ -3096,8 +3100,6 @@ LookupConnection (Coord X, Coord Y, bool AndDraw, Coord Range, int flag,
    */
   ListStart (type, ptr1, ptr2, ptr3, flag);
   DoIt (flag, AndRats, AndDraw);
-  if (store_undo)
-    IncrementUndoSerialNumber ();
   User = false;
 
   /* we are done */
@@ -3106,6 +3108,10 @@ LookupConnection (Coord X, Coord Y, bool AndDraw, Coord Range, int flag,
   if (AndDraw && Settings.RingBellWhenFinished)
     gui->beep ();
   FreeConnectionLookupMemory ();
+
+out:
+  if (store_undo)
+    IncrementUndoSerialNumber ();
 }
 
 /* ---------------------------------------------------------------------------
diff --git a/src/netclass.c b/src/netclass.c
index b81d0dc..d13d3f0 100644
--- a/src/netclass.c
+++ b/src/netclass.c
@@ -295,11 +295,14 @@ find_netname_for_object (int type, void *ptr1, void *ptr2, void *ptr3)
 {
   char *netname;
   int flag = CONNECTEDFLAG;
-  ClearFlagOnAllObjects (false /*AndDraw*/, flag, true /* store_undo */);
+  bool changed = false;
+
+  changed |= ClearFlagOnAllObjects (false /*AndDraw*/, flag, true /* store_undo */);
   /* Find a pin / pad which is connected, and grab the netname from this */
-  LookupObject (type, ptr1, ptr2, ptr3, false /*AndDraw*/, flag, true /*AndRats*/, true /*store_undo*/);
+  LookupObject (type, ptr1, ptr2, ptr3, false /*AndDraw*/, flag, true /*AndRats*/, true /*store_undo*/); /* This call bumps the undo serial number */
   netname = find_first_flagged_netname (flag);
-  Undo (true);
+//  if (changed)
+    Undo (true);
   return netname;
 }
 
@@ -308,11 +311,14 @@ find_netname_at_xy (LayerType *layer, Coord x, Coord y)
 {
   char *netname;
   int flag = CONNECTEDFLAG;
-  ClearFlagOnAllObjects (false /*AndDraw*/, flag, true /*store_undo*/);
+  bool changed = false;
+
+  changed |= ClearFlagOnAllObjects (false /*AndDraw*/, flag, true /*store_undo*/);
   /* Find a pin / pad which is connected, and grab the netname from this */
-  LookupConnection (x, y, false /*AndDraw*/, 0 /*Range*/, flag, true /*AndRats*/, true /*store_undo*/);
+  LookupConnection (x, y, false /*AndDraw*/, 0 /*Range*/, flag, true /*AndRats*/, true /*store_undo*/); /* This call bumps the undo serial number */
   netname = find_first_flagged_netname (flag);
-  Undo (true);
+//  if (changed)
+    Undo (true);
   return netname;
 }
